### PR TITLE
[quant][pt2] Fix conv bn decomposition in cuda mode

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108809

Summary: If the user calls model.cuda() before exporting
with `capture_pre_autograd_graph`, they get `cudnn_batch_norm`
instead of `native_batch_norm`, which is the op in cpu mode.
This commit makes both cases produce `native_batch_norm`
to be more consistent, otherwise QAT fusion would not happen.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_fusion_cuda
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_relu_fusion_cuda

Reviewers: jerryzh168, kimishpatel, bdhirsh

Subscribers: jerryzh168, kimishpatel, bdhirsh

Differential Revision: [D49472927](https://our.internmc.facebook.com/intern/diff/D49472927)

### CODE REVIEW

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py [hunk 1]
```diff
@@ -352,11 +352,6 @@ def _annotate_conv2d_bn(
         if len(conv_partition.output_nodes) > 1:
             raise ValueError("conv partition has more than one output node")
         conv_node = conv_partition.output_nodes[0]
-        conv_node_users = list(conv_node.users.keys())
-        if len(conv_node_users) > 1:
-            raise ValueError(
-                "Conv node must be consumed by BN only for it to be fusable."
```
Comment (bdhirsh) [inline torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:358]: is this change related to the rest of the PR?

**Review by bdhirsh — APPROVED** [review]
the cudnn_batch_norm decompositon change lgtm! Not sure about the change in `xnnpack_quantizer_utils.py`

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:358]: Yeah this check was added for the cpu graph, but the cuda graph has another user

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1782,6 +1782,7 @@ def nop_decomposition(x):
 # Also register to the Autograd dispatch key, so this decomp can run above autograd.
 # native_batch_norm needs to decompose into other ops before autograd.
 @aten.cudnn_batch_norm.default.py_impl(DispatchKey.Autograd)
+@aten.cudnn_batch_norm.default.py_impl(DispatchKey.CompositeImplicitAutograd)
```
Comment (kimishpatel) [inline torch/_decomp/decompositions.py:1785]: Wait @bdhirsh is this a safe thing to do? Now even in eager mode, you will run through this implementation, right? cudnn_bn has it own fw and bw impl https://fburl.com/code/8t71v9nj

**Comment by kimishpatel** [comment]
I am little confused. @bdhirsh does pre-dispatch export, run composite implicit decomps?

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (bdhirsh) [reply torch/_decomp/decompositions.py:1785]: @kimishpatel so what's going on here is that we are registering this decomposition to the **python dispatcher** (using `.py_impl`). The python dispatcher is opt-in, and can be turned on using `with enable_python_dispatcher()`.

In eager mode we do not turn it on, but during compilation in PT2 we always turn it on (this has been very convenient for changes decomps at the autograd level that have been painful for torch.compile, without having to worry about BC-breaking eager mode).

> is this a safe thing to do

Based on everything I said, a very reasonable q is "ok, but will this change the result of batch norm in eager vs. compile?".

We're relying here on the fact that we expect `torch.batch_norm` to give (nearly) the same output, regardless of whether or not cudnn was used. cudnn won't give bitwise equivalent outputs compared to our native batchnorm kernel, but compile already doesn't promise to be bitwise equivalent to eager mode - so we're basically saying here that "even if we would have run cudnn if the user called batch norm in eager mode, we will skip cudnn and use inductor if you are using compile".

**Comment by bdhirsh** [comment]
@kimishpatel I think the POR for pre_dispatch export is basically:

(1) pre_dispatch export will **not** decompose CompositeImplicitAutograd ops

(2) however, there are a few CompositeImplicit decomps that are required to run in order for functionalization to run properly, because they correspond to maybe-aliasing and maybe-mutating ops. There are ~6-7 ops in this list, and `cuddn_batch_norm` is one of them

@andrewor14 wanted to add these decomps sooner rather than later, so that we'll be less BC-breaking to existing pre_dispatch export users once we turn on pre-dispatch functionalization

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (bdhirsh) [reply torch/_decomp/decompositions.py:1785]: Hmm. I don't know what inductor issue you ran into, but one thing to mention is that if you want the decomp to run **before** autograd runs, then the decomp needs to be autograd-preserving.


For export though, it probably doesn't matter if this decomp runs before/after autograd, if you only care about inference.

Comment (kimishpatel) [reply torch/_decomp/decompositions.py:1785]: @larryliu0820 what was the inductor issue you ran into?

Comment (larryliu0820) [reply torch/_decomp/decompositions.py:1785]: Here: https://github.com/pytorch/pytorch/actions/runs/6018775992/job/16327865852

The PR commit is https://github.com/pytorch/pytorch/pull/108180/commits/e9818631222a7b595b4fbddf43cb80c6bf7687a4

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:358]: @andrewor14 you mean the cuda graph, after the bn decomp is added (in this PR), is still different from cpu graph?

**Comment by jerryzh168** [comment]
@andrewor14 I think the question here is, since native_bn and cudnn_bn doesn't have the same numerics, what if there are models just require to use cudnn_bn impl

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:358]: Yeah, it's just one op (forgot the name, something to do with memory format) but it's dead code and gets eliminated later anyway

**Comment by andrewor14** [comment]
Just wanted to give a quick update on this PR. There are a few moving pieces here:

(1) **Batch norm op is specialized on the device (cpu/cuda) during export.** Before this PR, we get `_native_batch_norm_legit` if the user exports with `model.cpu()`, and `cudnn_batch_norm` if the user exports with `model.cuda()`. We found that a lot of user training code only calls `model.cuda()` *after* export, meaning they would still get the `_native_batch_norm_legit` version in their graph.

(2) **cudnn_batch_norm has better numerics.** We found that `_native_batch_norm_legit` gave worse numerics than `cudnn_batch_norm` on cuda when performing QAT on mobilenetv2. We also confirmed this in a small experiment comparing the output of an exported mobilenetv2 vs a non-exported one (outside of quantization), and found that `_native_batch_norm_legit` output diverged significantly while `cudnn_batch_norm` output was identical to the baseline.

(3) **cudnn_batch_norm is not compatible with the long term.** `cudnn_batch_norm` as it is today does not have a functional variant, and so is incompatible with `torch.export` in the long run. There are a few options to work around this: we could either add the `cudnn_batch_norm_no_training` and `cudnn_batch_norm_functional` variants, or consolidate the batch norm ops such that we get a single op that is not specialized to a particular backend yet. The latter seems more attractive.

**Comment by andrewor14** [comment]
> @andrewor14 I think the question here is, since native_bn and cudnn_bn doesn't have the same numerics, what if there are models just require to use cudnn_bn impl

My opinion is we should still land this PR, so we'll end up with a single `_native_batch_norm_legit` op that QAT needs to handle. Users who need better numerics can manually swap this op out with `cudnn_batch_norm` after prepare. This is only a temporary solution until we have something better (like batch norm consolidation).

The alternative is to make QAT handle both ops instead of landing this PR. Right now exporting with `model.cuda()` is broken because we silently don't perform QAT fusion when we see `cudnn_batch_norm`. However, I'm less inclined towards this approach because it's more implicit; users will get good accuracies if they happen to call `model.cuda()` before export, but will get poor accuracies if they export first then call `model.cuda()`. Also it's more complexity QAT and quantizers need to handle.

**Comment by jerryzh168** [comment]
do you know why cudnn_batch_norm has better accuracy? I think having a single op sounds better but we also do not want to sacrifice the QAT accuracy, so we have to provide alternatives if that is required.

**Comment by kimishpatel** [comment]
@andrewor14 I want to highlight a couple of things

> (2) **cudnn_batch_norm has better numerics.** We found that `_native_batch_norm_legit` gave worse numerics than `cudnn_batch_norm` on cuda when performing QAT on mobilenetv2. We also confirmed this in a small experiment comparing the output of an exported mobilenetv2 vs a non-exported one (outside of quantization), and found that `_native_batch_norm_legit` output diverged significantly while `cudnn_batch_norm` output was identical to the baseline.

This seems concerning to me. While different op implementations can certainly diverge in numerics, I want to understand what is the meaning of the diverge a lot, here. And why cudnn implementation is considered to have better numerics? We should really address this because this has more serious implications on model training in general.

> My opinion is we should still land this PR, so we'll end up with a single _native_batch_norm_legit op that QAT needs to handle. Users who need better numerics can manually swap this op out with cudnn_batch_norm after prepare. This is only a temporary solution until we have something better (like batch norm consolidation).

Is it possible to have the same set of ops regardless of cpu vs. cuda transformations on model, before export? I think thats what we should really aim for. Can we figure out the work for this.

In the meantime, I would rather prefer us handling fusion for cudnn variant as well and have quantizer deal with this too. Reason is that QAT users will often enough want do `.cuda` on the model. That is likely more prominent usecase. (although I would like that .cpu/.cuda work on exported model as well. they shouldnt be generating two different opset, but this is likely gonna require more discussion)

**Comment by andrewor14** [comment]
Ok, after latest discussions with @jerryzh168, @kimishpatel, and @bdhirsh, we aligned on the following short term flow:
```python
# model can be in cpu or cuda mode
model = capture_pre_autograd_graph(model, example_inputs)
# at this point the graph contains _native_batch_norm_legit only
model = prepare_qat_pt2e(model, quantizer)
# some util function, exact name and location TBD
_swap_batch_norm_to_cudnn(model)
# at this point the graph contains cudnn_batch_norm only
train(model)
```
The above will be true regardless of whether the model was in cpu or cuda mode before exporting. This allows quantization to only deal with one op `_native_batch_norm_legit`, the one that is more consistent with the long term. In the short term, to get better numerics, the user can optionally call a util method that swaps this op with `cudnn_batch_norm` *after* prepare. The short term flow described here requires landing the changes in PR.

In the longer term we plan to get rid of this extra `_swap_batch_norm_to_cudnn` call by consolidating the batch norm ops, such that `_native_batch_norm_legit` will call the right kernel depending on whether cudnn is installed. This makes QAT give good numerics out of the box.

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49472927).

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (bdhirsh) [reply torch/_decomp/decompositions.py:1786]: Oh that error is fixable - you probably just need to talk to some folks from inductor (maybe Elias). Inductor is complaining because there is now a fallback *and* a decomposition defined for that flash attention op, so inductor does'nt know which one to use.

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:339]: I think technically this check can't be removed since otherwise the fusion will not be valid?

**Comment by andrewor14** [comment]
Closing this for now because it's a risky change for non-quantization users of `capture_pre_autograd_graph` and `torch._dynamo.export(pre_dispatch=False)`, who would not have access to the cudnn kernel anymore

**Comment by andrewor14** [comment]
Note: https://github.com/pytorch/pytorch/pull/109908 provides an alternative solution to this PR. There we support both `cudnn_batch_norm` and `_native_batch_norm_legit` and do not attempt to "consolidate" these two ops for QAT (and pre-dispatch export in general)
